pkg/keyspace/tso_keyspace_group.go

```diff
@@ -286,33 +290,70 @@ func (m *GroupManager) GetKeyspaceGroupByID(id uint32) (*endpoint.KeyspaceGroup,
 // DeleteKeyspaceGroupByID deletes the keyspace group by ID.
 func (m *GroupManager) DeleteKeyspaceGroupByID(id uint32) (*endpoint.KeyspaceGroup, error) {
 	var (
-		kg  *endpoint.KeyspaceGroup
-		err error
+		kg           *endpoint.KeyspaceGroup
+		err          error
+		moreToDelete = true
 	)
 
 	m.Lock()
 	defer m.Unlock()
-	if err := m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
-		kg, err = m.store.LoadKeyspaceGroup(txn, id)
+	for moreToDelete {
+		err = m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
+			kg, err = m.store.LoadKeyspaceGroup(txn, id)
+			if err != nil {
+				return err
+			}
+			if kg == nil {
+				return nil
+			}
+			if kg.IsSplitting() {
+				return ErrKeyspaceGroupInSplit
+			}
+
+			var (
+				keyspaceIDsToUnlock = make([]uint32, 0, maxTxnOps)
+				deleteKeyspaceCount = 0
+			)
+			defer func() {
+				for _, id := range keyspaceIDsToUnlock {
+					m.km.metaLock.Unlock(id)
+				}
+			}()
+			for _, id := range kg.Keyspaces {
+				if deleteKeyspaceCount >= maxTxnOps {
+					moreToDelete = true
+					kg.Keyspaces = kg.Keyspaces[deleteKeyspaceCount:]
+					return m.store.SaveKeyspaceGroup(txn, kg)
+				}
+				deleteKeyspaceCount++
+
+				m.km.metaLock.Lock(id)
+				keyspaceIDsToUnlock = append(keyspaceIDsToUnlock, id)
+
+				ks, err := m.km.store.LoadKeyspaceMeta(txn, id)
+				if err != nil {
+					return err
+				}
+				if ks.Config == nil {
+					continue
+				}
+				delete(ks.Config, TSOKeyspaceGroupIDKey)
+
+				err = m.km.store.SaveKeyspaceMeta(txn, ks)
+				if err != nil {
+					log.Error("failed to save keyspace meta during removing keyspace group", zap.Uint32("keyspace-id", ks.Id), zap.Error(err))
+					return err
+				}
+			}
+			moreToDelete = false
+			return m.store.DeleteKeyspaceGroup(txn, id)
+		})
 		if err != nil {
-			return err
-		}
-		if kg == nil {
-			return nil
-		}
-		if kg.IsSplitting() {
-			return ErrKeyspaceGroupInSplit
+			return nil, err
 		}
-		return m.store.DeleteKeyspaceGroup(txn, id)
-	}); err != nil {
-		return nil, err
+		m.groups[endpoint.StringUserKind(kg.UserKind)].Put(kg)
```

Comment:
Is there any keyspace meta in memory? If yes, we need to update it too -- just like we update keyspace group meta in memory here.